README.md
beeswarm
========

[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/beeswarm)](https://cran.r-project.org/package=beeswarm)
[![CRAN_Downloads_Badge](http://cranlogs.r-pkg.org/badges/beeswarm)](https://cran.r-project.org/package=beeswarm)

An R package implementing bee swarm plots


Installation
------------

You can install the latest release on CRAN like this:

	install.packages("beeswarm")


Or you can install the latest development version from GitHub like this:

	## install.packages("devtools")
	devtools::install_github("aroneklund/beeswarm")


Related works
-------------

* [**ggbeeswarm**](https://cran.r-project.org/package=ggbeeswarm) 
bee swarm and related plots in the ggplot2 framework
* [**pybeeswarm**](https://github.com/mgymrek/pybeeswarm)
bee swarm plots in Python


It's also known as a dot plot
-----------------------------

Leland Wilkinson (1999). Dot plots. The American Statistician. 53(3):276-281. 
[PDF](https://www.cs.uic.edu/~wilkinson/Publications/dotplots.pdf)

It seems to be the same basic point layout algorithm, except that Wilkinson suggests an additional smoothing step that the beeswarm package does not do.


NEWS
beeswarm NEWS

Changes in development version

- swarmx() and swarmy() can handle infinite and NA values (GitHub Issue #19)


Changes in version 0.4.0 (2021-05-07)

- Thanks to James Trimble (now a package author) for the following:
- The "method" argument has a new option "compactswarm", which gives a swarm with a more dense point layout.
- New argument "fast" (default: TRUE) does point layout with faster C code.
- Thus, the package now requires compilation.


Changes in version 0.3.1 (2021-03-06)

- Fixed bug where non-varying input (a single value, or a single value repeated) in non-swarm methods would yield NAs (thanks raredd). (GitHub Issue #8)
- Updated email and URL in DESCRIPTION.


Changes in version 0.3.0 (2017-05-30)

- Added new argument "pwcex" for adjusting plotting character size of individual points. (GitHub request #3 by alexpreynolds)


Changes in version 0.2.3 (2016-04-25)

- Fixed a bug in which specifying non-default "corral" and "side" arguments would result in unexpected results. (Issue #2)


Changes in version 0.2.2 (2015-11-25)

- Check that glim, dlim, xlim, ylim have length 2, thus avoiding confusing errors.
- New argument "axes" (previously, trying to specify "axes" in ... would cause an error).


Changes in version 0.2.1 (2015-08-26)

- Corrected NAMESPACE and DESCRIPTION to properly indicate Imports and to remove unnecessary Depends.


Changes in version 0.2.0 (2015-05-04)

- New arguments "side" and "priority" for beeswarm, swarmx, and swarmy. "side" allows swarms to grow on only one side rather than both sides. "priority" controls the order in which swarms are built, thus affecting the appearance of the plot. Thanks to Jon Lake and Brad Stiritz for suggestions.


Changes in version 0.1.8 (2015-04-30)

- Added "Depends: stats" to DESCRIPTION and removed "require(stats)" calls from code in order to achieve a clean package check


Changes in version 0.1.7 (2014-08-05)

- ".calculateSwarm" now performs the swarm point layout more efficiently


Changes in version 0.1.6 (2013-09-18)

- The argument "labels" now gets recycled.
- "labels = NULL" is now the same as missing "labels"; i.e. labels are inferred from data.
- There is now a "corralWidth" argument to control the size of corrals, if corrals are used.


Changes in version 0.1.5 (2012-05-07)

- Minor adjustments.


Changes in version 0.1.4 (2012-05-03)

- Minor adjustments.


Changes in version 0.1.3 (2012-03-22)

- New function "bxplot" as a minimal version of "boxplot", intended mainly as a way to add quantile lines to a beeswarm plot.
- "beeswarm" has a new argument "corral" to control the spread of swarms into adjacent groups.


Changes in version 0.1.2 (2012-03-14)

- Documentation fixes
- Added a NAMESPACE file
- The formula interface now splits as expected for formulas such as: x ~ y * z


Changes in version 0.1.1 (2011-08-04)

- Documentation fixes.
- Adjusted "beeswarm.default" to work when x is a simple numeric vector.
- Removed "beeswarm.numeric" .


Changes in version 0.1.0 (2011-08-03)

- In "beeswarm", defaults for "col", "pch" are now taken from "par", and "bg" defaults to NA, and "pwbg" defaults to NULL.
- Renamed "smile" method; now it is called "swarm"
- Changed default method to "swarm"
- Removed function "smile"
- New functions "swarmx" and "swarmy" for easily adding swarm-ed points to a plot.
- Fixed bug affecting swarming when par('cex') was not 1.
- Fixed bug reversing xlim and ylim on horizontal plots.

tests/beeswarm-test.R
library(beeswarm)

test_swarms <- function(x) {
  for (na_positions in list(integer(0), c(2L,4L,6L,8L))) {
    x[na_positions] <- NA
    for (compact in c(FALSE, TRUE)) {
      for (priority in c("ascending", "descending", "density", "random", "none")) {
        for (side in -1:1) {
          print(na_positions)
          print(compact)
          print(priority)
          print(side)
          # compare R and C versions of swarmy and swarmx
          set.seed(1)
          y1 <- swarmy(x, numeric(length(x)), compact=compact, side=side, priority=priority, fast=TRUE)$y
          set.seed(1)
          y2 <- swarmy(x, numeric(length(x)), compact=compact, side=side, priority=priority, fast=FALSE)$y
          stopifnot(all.equal(y1, y2))
          # No values should be NA after swarm, regardless of NA in the other direction
          stopifnot(!any(is.na(y1)))

          set.seed(1)
          x1 <- swarmx(numeric(length(x)), x, compact=compact, side=side, priority=priority, fast=TRUE)$x
          set.seed(1)
          x2 <- swarmx(numeric(length(x)), x, compact=compact, side=side, priority=priority, fast=FALSE)$x
          stopifnot(all.equal(x1, x2))
          # No values should be NA after swarm, regardless of NA in the other direction
          stopifnot(!any(is.na(x1)))
        }
      }
    }
  }
}

set.seed(1)
test_swarms(rnorm(250) / 100)
test_swarms(numeric(10))
test_swarms(1:10)

# Ensure no error with infinite values (GitHub issue #19)
stopifnot(any(swarmx(x = rep(1, 5), y = c(rep(-Inf, 3), 0, 1))$x > 1))
stopifnot(any(swarmy(x = c(rep(-Inf, 3), 0, 1), y = rep(1, 5))$y > 1))

# Confirm that xToFinite() works as expected
stopifnot(xToFinite(NA, 10) == -200)
stopifnot(xToFinite(Inf, 10) == 100)
stopifnot(xToFinite(-Inf, 10) == -100)
stopifnot(xToFinite(c(-Inf, 1), 10) == c(-99, 1))
stopifnot(xToFinite(c(-Inf, Inf, NA, 1), 10) == c(-99, 101, -199, 1))


R/bxplot.R
# bxplot.R
#
# Aron Charles Eklund
##
# A part of the "beeswarm" R package
# 



bxplot <- function (x, ...) 
  UseMethod("bxplot")

bxplot.default <- function(x, probs = c(0.25, 0.5, 0.75),
    vertical = TRUE, horizontal = !vertical, add = FALSE,
    col = par("col"), lty = par("lty"), lwd = NULL, 
    at = NULL, width = 0.75, ...) {
  if(is.numeric(x)) {
    x <- list(x)
  }
  n <- length(x)
  n.probs <- length(probs)
  if(is.null(lwd)) {  ## default is a thick line at the median
    lwd <- rep(par('lwd'), length.out = n.probs)
    if(0.5 %in% probs) lwd[probs == 0.5] <- par('lwd') * 3
  }
  y <- lapply(x, quantile, probs = probs, na.rm = TRUE)
  if(is.null(at)) at <- 1:n
  if(!add) {
    boxplot(x, horizontal = horizontal, at = at, 
      pars = list(whisklty = 0, staplelty = 0, outpch = NA, 
        boxlty = 0, medlty = 0), ...)
  }
  hw <- width / 2  # half-width
  if(horizontal) {
    for (i in 1:n) {
      segments(y0 = at[i] - hw, y1 = at[i] + hw, x0 = y[[i]], 
        col = col, lwd = lwd, lty = lty)
    }
  } else {
    for (i in 1:n) {
      segments(x0 = at[i] - hw, x1 = at[i] + hw, y0 = y[[i]], 
        col = col, lwd = lwd, lty = lty)
    }
  }
}  

bxplot.formula <- function (formula, data = NULL, ..., subset, na.action = NULL) 
{
    if (missing(formula) || (length(formula) != 3L)) 
        stop("'formula' missing or incorrect")
    m <- match.call(expand.dots = FALSE)
    if (is.matrix(eval(m$data, parent.frame()))) 
        m$data <- as.data.frame(data)
    m$... <- NULL
    m$na.action <- na.action
    m[[1L]] <- as.name("model.frame")
    mf <- eval(m, parent.frame())
    response <- attr(attr(mf, "terms"), "response")
    bxplot(split(mf[[response]], mf[-response]), ...)
}




R/beeswarm.R
# beeswarm.R
#
# Aron Charles Eklund
#
# A part of the "beeswarm" R package
#


beeswarm <- function (x, ...)
  UseMethod("beeswarm")


## here x should be a list or data.frame or numeric
beeswarm.default <- function(x,
    method = c("swarm", "compactswarm", "center", "hex", "square"),
    vertical = TRUE, horizontal = !vertical,
    cex = 1, spacing = 1, breaks = NULL,
    labels, at = NULL,
    corral = c("none", "gutter", "wrap", "random", "omit"),
    corralWidth, side = 0L,
    priority = c("ascending", "descending", "density", "random", "none"),
    fast = TRUE,
    pch = par("pch"), col = par("col"), bg = NA,
    pwpch = NULL, pwcol = NULL, pwbg = NULL, pwcex = NULL,
    do.plot = TRUE, add = FALSE, axes = TRUE, log = FALSE,
    xlim = NULL, ylim = NULL, dlim = NULL, glim = NULL,
    xlab = NULL, ylab = NULL, dlab = "", glab = "",
    ...) {

  method <- match.arg(method)
  corral <- match.arg(corral)
  priority <- match.arg(priority)
  if(length(cex) > 1) {
    stop('the parameter "cex" must have length 1')
  }
  stopifnot(side %in% -1:1)
  if(is.numeric(x)) {
    x <- list(x)
  }
  n.groups <- length(x)

  #### Resolve group labels
  if(missing(labels) || is.null(labels)) {
    if(is.null(names(x))) {
      if(n.groups == 1) {
        labels <- NA
      } else {
        labels <- 1:n.groups
      }
    } else {
      labels <- names(x)
    }
  } else {
    labels <- rep(labels, length.out = n.groups)
  }

  if (is.null(at))
    at <- 1:n.groups
  else if (length(at) != n.groups)
    stop(gettextf("'at' must have length equal to %d, the number of groups",
      n.groups), domain = NA)

  if (is.null(dlab))
     dlab <- deparse(substitute(x))

  ## this function returns a "group" vector, to complement "unlist"
  unlistGroup <- function(x, nms = names(x)) rep(nms, sapply(x, length))

  x.val <- unlist(x)
  x.gp <- unlistGroup(x, nms = labels)
  if((range(x.val, finite = TRUE)[1] <= 0) && log)
    warning('values <= 0 omitted from logarithmic plot')

  n.obs <- length(x.val)
  n.obs.per.group <- lengths(x)

  #### Resolve xlim, ylim, dlim, xlab, ylab
  if(is.null(dlim)) {
      if(log) {
        dlim <- 10 ^ (extendrange(log10(x.val[x.val > 0])))
      } else {
        dlim <- extendrange(x.val, f = 0.01)
      }
  } else if (length(dlim) != 2) {
    stop ("'dlim' must have length 2")
  }
  if(is.null(glim)) {
    glim <- c(min(at) - 0.5, max(at) + 0.5)
  } else if (length(glim) != 2) {
    stop ("'glim' must have length 2")
  }
  if(horizontal) {        ## plot is horizontal
    if(is.null(ylim))
      ylim <- glim
    if(is.null(xlim)) {
      xlim <- dlim
    } else {
      dlim <- xlim
    }
    if (is.null(xlab))
      xlab <- dlab
    if (is.null(ylab))
      ylab <- glab
  } else {                ## plot is vertical
    if(is.null(xlim))
      xlim <- glim
    if(is.null(ylim)) {
      ylim <- dlim
    } else {
      dlim <- ylim
    }
    if (is.null(ylab))
      ylab <- dlab
    if (is.null(xlab))
      xlab <- glab
  }
  if(length(xlim) != 2)
    stop ("'xlim' must have length 2")
  if(length(ylim) != 2)
    stop ("'ylim' must have length 2")

  #### Resolve plotting characters and colors
  if(is.null(pwpch)) {
    pch.out <- unlistGroup(x, nms = rep(pch, length.out = n.groups))
  } else {
    if(is.list(pwpch)) {
      names(pwpch) <- names(x)
      stopifnot(all(sapply(pwpch, length) == n.obs.per.group))
      pch.out <- unlist(pwpch)
    } else {
      pch.out <- pwpch
    }
  }
  stopifnot(length(pch.out) == n.obs)

  if(is.null(pwcol)) {
    col.out <- unlistGroup(x, nms = rep(col, length.out = n.groups))
  } else {
    if(is.list(pwcol)) {
      names(pwcol) <- names(x)
      stopifnot(all(sapply(pwcol, length) == n.obs.per.group))
      col.out <- unlist(pwcol)
    } else {
      col.out <- pwcol
    }
  }
  stopifnot(length(col.out) == n.obs)

  if(is.null(pwbg)) {
    bg.out <- unlistGroup(x, nms = rep(bg, length.out = n.groups))
  } else {
    if(is.list(pwbg)) {
      names(pwbg) <- names(x)
      stopifnot(all(sapply(pwbg, length) == n.obs.per.group))
      bg.out <- unlist(pwbg)
    } else {
      bg.out <- pwbg
    }
  }
  stopifnot(length(bg.out) == n.obs)

   if(is.null(pwcex)) {
    cex.out <- unlistGroup(x, nms = rep(1, length.out = n.groups))
  } else {
    if(is.list(pwcex)) {
      names(pwcex) <- names(x)
      stopifnot(all(sapply(pwcex, length) == n.obs.per.group))
      cex.out <- unlist(pwcex)
    } else {
      cex.out <- pwcex
    }
  }
  stopifnot(length(cex.out) == n.obs)

  #### Set up the plot
  if(do.plot & !add) {
    plot(xlim, ylim,
      type = 'n', axes = FALSE,
      log = ifelse(log, ifelse(horizontal, 'x', 'y'), ''),
      xlab = xlab, ylab = ylab, ...)
  }

  #### Calculate the size of a plotting character along group- or data-axis
  sizeMultiplier <- par('cex') * cex * spacing
  if(horizontal) {
    size.g <- yinch(0.08, warn.log = FALSE) * sizeMultiplier
    size.d <- xinch(0.08, warn.log = FALSE) * sizeMultiplier
  } else {    # vertical
    size.g <- xinch(0.08, warn.log = FALSE) * sizeMultiplier
    size.d <- yinch(0.08, warn.log = FALSE) * sizeMultiplier
  }

  ##### Calculate point positions g.pos, d.pos
  if(method %in% c('swarm', 'compactswarm')) {
    compact <- method == 'compactswarm'
    if(horizontal) {
      g.offset <- lapply(x, function(a) swarmy(x = a, y = rep(0, length(a)),
          cex = sizeMultiplier, side = side, priority = priority,
          fast = fast, compact = compact)$y)
    } else {
      g.offset <- lapply(x, function(a) swarmx(x = rep(0, length(a)), y = a,
          cex = sizeMultiplier, side = side, priority = priority,
          fast = fast, compact = compact)$x)
    }
    d.pos <- x
  } else {          ####   non-swarm methods
    ##### first determine positions along the data axis
      if(method == 'hex') size.d <- size.d * sqrt(3) / 2
      if(log) {              ## if data axis IS on a log scale
        if(is.null(breaks))
          breaks <- 10 ^ seq(log10(dlim[1]), log10(dlim[2]) + size.d, by = size.d)
        if(length(breaks) == 1 && is.na(breaks[1])) {
          d.index <- x
          d.pos <- x
        } else {
          mids <- 10 ^ ((log10(head(breaks, -1)) + log10(tail(breaks, -1))) / 2)
          d.index <- lapply(x, cut, breaks = breaks, labels = FALSE, include.lowest = TRUE)
          d.pos <- lapply(d.index, function(a) mids[a])
        }
      } else {               ## if data axis is NOT on a log scale
        if(is.null(breaks))
          breaks <- seq(dlim[1], dlim[2] + size.d, by = size.d)
        if(length(breaks) == 1 && is.na(breaks[1])) {
          d.index <- x
          d.pos <- x
        } else {
          mids <- (head(breaks, -1) + tail(breaks, -1)) / 2
          d.index <- lapply(x, cut, breaks = breaks, labels = FALSE, include.lowest = TRUE)
          d.pos <- lapply(d.index, function(a) mids[a])
        }
      }
    ##### now determine positions along the group axis
      x.index <- lapply(d.index, function(v) {
        if(length(na.omit(v)) == 0)
          return(v)
        v.s <- lapply(split(v, v), seq_along)
        if(method %in% c('center', 'square') && side == -1)
          v.s <- lapply(v.s, function(a) a - max(a))
        else if(method %in% c('center', 'square') && side == 1)
          v.s <- lapply(v.s, function(a) a - 1)
        else if(method == 'center')
          v.s <- lapply(v.s, function(a) a - mean(a))
        else if(method == 'square')
          v.s <- lapply(v.s, function(a) a - floor(mean(a)))
        else if(method == 'hex') {
          odd.row <- (as.numeric(names(v.s)) %% 2) == 1
          if(side == 0) {
            v.s[ odd.row] <- lapply(v.s[ odd.row], function(a) a - floor(mean(a)) - 0.25)
            v.s[!odd.row] <- lapply(v.s[!odd.row], function(a) a - ceiling(mean(a)) + 0.25)
          } else if(side == -1) {
            v.s[ odd.row] <- lapply(v.s[ odd.row], function(a) a - max(a))
            v.s[!odd.row] <- lapply(v.s[!odd.row], function(a) a - max(a) - 0.5)
          } else if(side ==  1) {
            v.s[ odd.row] <- lapply(v.s[ odd.row], function(a) a - 1)
            v.s[!odd.row] <- lapply(v.s[!odd.row], function(a) a - 0.5)
          }
        }
        unsplit(v.s, v)
      })

      g.offset <- lapply(1:n.groups, function(i) x.index[[i]] * size.g)
  }                   ###### end of non-swarm methods

  ##### now check for runaway points (if "corral" has been set)
  if(corral != 'none') {
    if(missing(corralWidth)) {
      if(n.groups > 1) {
        corralWidth <- min(at[-1] - at[-n.groups]) - (2 * size.g)
      } else {
        corralWidth <- 2 * (min(diff(c(par('usr')[1], at, par('usr')[2]))) - size.g)
      }
    } else {
      stopifnot(length(corralWidth) == 1)
      stopifnot(corralWidth > 0)
    }
    corralLo <- (side - 1) * corralWidth / 2
    corralHi <- (side + 1) * corralWidth / 2
    if(corral == 'gutter') {
      g.offset <- lapply(g.offset, function(zz) pmin(corralHi, pmax(corralLo, zz)))
    }
    if(corral == 'wrap') {
      if(side == -1) { ## special case with side=-1: reverse the corral to avoid artifacts at zero
        g.offset <- lapply(g.offset, function(zz) corralHi - ((corralHi - zz) %% corralWidth))
      } else {
        g.offset <- lapply(g.offset, function(zz) ((zz - corralLo) %% corralWidth) + corralLo)
      }
    }
    if(corral == 'random') {
      g.offset <- lapply(g.offset, function(zz) ifelse(zz > corralHi | zz < corralLo, yes = runif(length(zz), corralLo, corralHi), no = zz))
    }
    if(corral == 'omit') {
      g.offset <- lapply(g.offset, function(zz) ifelse(zz > corralHi | zz < corralLo, yes = NA, no = zz))
    }
  }

  g.pos <- lapply(1:n.groups, function(i) at[i] + g.offset[[i]])

  out <- data.frame(x = unlist(g.pos), y = unlist(d.pos),
                    pch = pch.out, col = col.out, bg = bg.out, cex = cex * cex.out,
                    x.orig = x.gp, y.orig = x.val,
                    stringsAsFactors = FALSE)

  if(do.plot) {
    if(horizontal) {     ## plot is horizontal
      points(out$y, out$x, pch = out$pch, col = out$col, bg = out$bg, cex = out$cex)
      if(axes & !add) {
        axis(1, ...)
        axis(2, at = at, labels = labels, tick = FALSE, ...)
        box(...)
      }
    } else {             ## plot is vertical
      points(out$x, out$y, pch = out$pch, col = out$col, bg = out$bg, cex = out$cex)
      if(axes & !add) {
        axis(2, ...)
        axis(1, at = at, labels = labels, tick = FALSE, ...)
        box(...)
      }
    }
  }
  invisible(out)
}



beeswarm.formula <- function (formula, data = NULL, subset, na.action = NULL,
    pwpch = NULL, pwcol = NULL, pwbg = NULL, pwcex = NULL, dlab, glab, ...)
{
    if (missing(formula) || (length(formula) != 3))
        stop("'formula' missing or incorrect")
    m <- match.call(expand.dots = FALSE)
    if (is.matrix(eval(m$data, parent.frame())))
        m$data <- as.data.frame(data)
    m$... <- NULL
    m$dlab <- NULL
    m$glab <- NULL
    m$na.action <- na.action
    m[[1]] <- as.name("model.frame")
    mf <- eval(m, parent.frame())
    response <- attr(attr(mf, "terms"), "response")
    if (missing(dlab))
        dlab <- names(mf)[response]
    if (missing(glab))
        glab <- as.character(formula)[3]
    f <- mf[-response]
    f <- f[names(f) %in% attr(attr(mf, "terms"), "term.labels")]
    if(!is.null(mf$'(pwpch)')) pwpch <- split(mf$'(pwpch)', f)
    if(!is.null(mf$'(pwcol)')) pwcol <- split(mf$'(pwcol)', f)
    if(!is.null(mf$'(pwbg)')) pwbg <- split(mf$'(pwbg)',f)
    if(!is.null(mf$'(pwcex)')) pwcex <- split(mf$'(pwcex)',f)
    beeswarm(split(mf[[response]], f),
      pwpch = pwpch, pwcol = pwcol, pwbg = pwbg, pwcex = pwcex,
      dlab = dlab, glab = glab, ...)
}


#### hidden function to do swarm layout
.calculateSwarm <- function(x, dsize, gsize, side = 0L, priority = "ascending") {
  if(length(x) == 0) return(numeric(0))
  stopifnot(side %in% -1:1)
  out <- data.frame(x = x / dsize, y = 0, index = seq_along(x))

  #### Determine the order in which points will be placed
  if(     priority == "ascending" ) { out <- out[order( out$x), ] } ## default "smile"
  else if(priority == "descending") { out <- out[order(-out$x), ] } ## frown
  else if(priority == "none") {  } ## do not reorder
  else if(priority == "density") {
  	dens.x <- density(out$x, na.rm = TRUE)  ## compute kernel density estimate
  	dens.interp <- approx(dens.x$x, dens.x$y, xout = out$x, rule = 2)  ## interpolated density
  	out <- out[order(-dens.interp$y), ]  ## arrange outward from densest areas
  }
  else if(priority == "random") {
	out <- out[sample(nrow(out)), ]
  }
  #### place the points
  if(nrow(out) > 1) {
    for (ii in 2:nrow(out)) {          ## we will place one point at a time
      xi <- out$x[ii]
       ## identify previously-placed points with potential to overlap the current point
      isPotOverlap <- (abs(xi - out$x) < 1) & (1:nrow(out) < ii)
      isPotOverlap[is.na(isPotOverlap)] <- FALSE
      if(any(isPotOverlap)) {
        pre.x <- out[isPotOverlap, 'x']
        pre.y <- out[isPotOverlap, 'y']
        poty.off <- sqrt(1 - ((xi - pre.x) ^ 2))         ## potential y offsets
        poty <- switch(side + 2,
          c(0, pre.y - poty.off),
          c(0, pre.y + poty.off, pre.y - poty.off),
          c(0, pre.y + poty.off)
        )
        poty.bad <- sapply(poty, function(y) {           ## check for overlaps
          any(((xi - pre.x) ^ 2 + (y - pre.y) ^ 2) < 0.999)
        })
        poty[poty.bad] <- Inf
        out$y[ii] <- poty[which.min(abs(poty))]
      } else {
        out$y[ii] <- 0
      }
    }
  }
  out[is.na(out$x), 'y'] <- NA        ## missing x values should have missing y values
  out$y[order(out$index)] * gsize
}


.calculateCompactSwarm <- function(x, dsize, gsize, side = 0L, priority = "ascending") {
  if(length(x) == 0) return(numeric(0))
  stopifnot(side %in% -1:1)

  ## out$y.low: best permitted position <= 0 for each point
  ## out$y.high: best permitted position >= 0 for each point
  ## out$y.best: best permitted position for each point (or Inf for placed points)
  ## out$y.placed: which points have been placed
  out <- data.frame(x = x / dsize, y = 0, index = seq_along(x), placed = FALSE,
                    y.low = 0, y.high = 0, y.best = 0)

  #### Determine the order in which points will be placed
  if(     priority == "ascending" ) { out <- out[order( out$x), ] } ## default "smile"
  else if(priority == "descending") { out <- out[order(-out$x), ] } ## frown
  else if(priority == "none") {  } ## do not reorder
  else if(priority == "density") {
    dens.x <- density(out$x, na.rm = TRUE)  ## compute kernel density estimate
    dens.interp <- approx(dens.x$x, dens.x$y, xout = out$x, rule = 2)  ## interpolated density
    out <- out[order(-dens.interp$y), ]  ## arrange outward from densest areas
  }
  else if(priority == "random") {
    out <- out[sample(nrow(out)), ]
  }
  #### place the points
  if(nrow(out) > 1) {
    for (iter in 1:nrow(out)) {          ## we will place one point at a time
      i <- which.min(abs(out$y.best))    ## Choose a point that can be placed
                                         ## close to non-data axis
      xi <- out$x[i]
      yi <- out$y[i] <- out$y.best[i]
      out$placed[i] <- TRUE
      out$y.best[i] <- Inf               ## Ensure it won't be chosen again
      xdiff = abs(xi - out$x)
      if(side == 0) {
        for (j in which(!out$placed & xdiff < 1)) {
          y.offset <- sqrt(1 - (xdiff[j] ^ 2))
          y.high <- out$y.high[j] <- max(out$y.high[j], yi + y.offset)
          y.low <- out$y.low[j] <- min(out$y.low[j], yi - y.offset)
          out$y.best[j] <- ifelse(-y.low < y.high, y.low, y.high)
        }
      } else if(side == 1) {
        for (j in which(!out$placed & xdiff < 1)) {
          y.offset <- sqrt(1 - (xdiff[j] ^ 2))
          out$y.best[j] <- out$y.high[j] <- max(out$y.high[j], yi + y.offset)
        }
      } else {
        for (j in which(!out$placed & xdiff < 1)) {
          y.offset <- sqrt(1 - (xdiff[j] ^ 2))
          out$y.best[j] <- out$y.low[j] <- min(out$y.low[j], yi - y.offset)
        }
      }
    }
  }
  out[is.na(out$x), 'y'] <- NA        ## missing x values should have missing y values
  out$y[order(out$index)] * gsize
}


.calculateSwarmUsingC <- function(x, dsize, gsize, side = 0L, priority = "ascending",
    compact = FALSE) {
  if(length(x) == 0) return(numeric(0))
  stopifnot(side %in% -1:1)

  out <- data.frame(x = x / dsize, index = seq_along(x))

  #### Determine the order in which points will be placed
  if(     priority == "ascending" ) { out <- out[order( out$x), ] } ## default "smile"
  else if(priority == "descending") { out <- out[order(-out$x), ] } ## frown
  else if(priority == "none") {  } ## do not reorder
  else if(priority == "density") {
    dens.x <- density(out$x, na.rm = TRUE)  ## compute kernel density estimate
    dens.interp <- approx(dens.x$x, dens.x$y, xout = out$x, rule = 2)  ## interpolated density
    out <- out[order(-dens.interp$y), ]  ## arrange outward from densest areas
  }
  else if(priority == "random") {
    out <- out[sample(nrow(out)), ]
  }

  out <- out[!is.na(out$x), ]
  n <- nrow(out)

  #### place the points
  result <- .C(C_calculateSwarm,
               x = as.double(out$x),
               n = n,
               compact = as.logical(compact),
               side = as.integer(side),
               placed = integer(n),         # used internally by C implementations
               workspace = numeric(n * 4),  # used internally by C implementations
               y = numeric(n))
  y <- rep(NA, length(x))
  y[out$index] <- result[[7]] * gsize
  y
}

#### handle NA, positive and negative infinity so that they are considered
#### categorically
xToFinite <- function(x, dsize) {
  maskInf <- is.infinite(x)
  maskNa <- is.na(x)

  values <- list()
  if (all(maskInf | maskNa)) {
    values$na <- -20*dsize
    values$negInf <- -10*dsize
    values$posInf <- 10*dsize
  } else {
    positions <- range(x[!maskInf], na.rm = TRUE)
    values$na <- positions[1] - 20*dsize
    values$negInf <- positions[1] - 10*dsize
    values$posInf <- positions[2] + 10*dsize
  }

  if (any(maskNa)) {
    x[maskNa] <- values$na
  }
  if (any(maskInf)) {
    maskPosInf <- is.infinite(x) & !is.na(x) & x > 0
    maskNegInf <- is.infinite(x) & !is.na(x) & x < 0
    x[maskPosInf] <- values$posInf
    x[maskNegInf] <- values$negInf
  }
  x
}

### jitter points horizontally
swarmx <- function(x, y,
    xsize = xinch(0.08, warn.log = FALSE),
    ysize = yinch(0.08, warn.log = FALSE),
    log = NULL, cex = par("cex"), side = 0L,
    priority = c("ascending", "descending", "density", "random", "none"),
    fast = TRUE,
    compact = FALSE) {
  priority <- match.arg(priority)
  if(is.null(log))
    log <- paste(ifelse(par('xlog'), 'x', ''), ifelse(par('ylog'), 'y', ''), sep = '')
  xlog <- 'x' %in% strsplit(log, NULL)[[1L]]
  ylog <- 'y' %in% strsplit(log, NULL)[[1L]]
  xy <- xy.coords(x = x, y = y, recycle = TRUE, log = log)
  stopifnot((length(unique(xy$x)) <= 1))
  if(xlog) xy$x <- log10(xy$x)
  if(ylog) xy$y <- log10(xy$y)

  dsize <- ysize * cex
  otherval <- xToFinite(xy$y, dsize = dsize)

  if (fast) {
    x.new <- xy$x + .calculateSwarmUsingC(otherval, dsize = dsize,
      gsize = xsize * cex, side = side, priority = priority, compact = compact)
  } else {
    swarmFn <- ifelse(compact, .calculateCompactSwarm, .calculateSwarm)
    x.new <- xy$x + swarmFn(otherval, dsize = dsize, gsize = xsize * cex,
      side = side, priority = priority)
  }
  out <- data.frame(x = x.new, y = y)
  if(xlog) out$x <- 10 ^ out$x
  out
}

### jitter points vertically
swarmy <- function(x, y,
    xsize = xinch(0.08, warn.log = FALSE),
    ysize = yinch(0.08, warn.log = FALSE),
    log = NULL, cex = par("cex"), side = 0L,
    priority = c("ascending", "descending", "density", "random", "none"),
    fast = TRUE,
    compact = FALSE) {
  priority <- match.arg(priority)
  if(is.null(log))
    log <- paste(ifelse(par('xlog'), 'x', ''), ifelse(par('ylog'), 'y', ''), sep = '')
  xlog <- 'x' %in% strsplit(log, NULL)[[1L]]
  ylog <- 'y' %in% strsplit(log, NULL)[[1L]]
  xy <- xy.coords(x = x, y = y, recycle = TRUE, log = log)
  stopifnot((length(unique(xy$y)) <= 1))
  if(xlog) xy$x <- log10(xy$x)
  if(ylog) xy$y <- log10(xy$y)

  dsize <- xsize * cex
  otherval <- xToFinite(xy$x, dsize = dsize)

  if (fast) {
    y.new <- xy$y + .calculateSwarmUsingC(otherval, dsize = dsize,
      gsize = ysize * cex, side = side, priority = priority, compact = compact)
  } else {
    swarmFn <- ifelse(compact, .calculateCompactSwarm, .calculateSwarm)
    y.new <- xy$y + swarmFn(otherval, dsize = dsize, gsize = ysize * cex,
      side = side, priority = priority)
  }
  out <- data.frame(x = x, y = y.new)
  if(ylog) out$y <- 10 ^ out$y
  out
}
